edb/server/protocol/binary.pyx

```diff
@@ -449,26 +452,68 @@ cdef class EdgeConnection(frontend.FrontendConnection):
                 f'authentication failed: cannot decode JWT'
             ) from None
 
-        namespace = "edgedb.server"
-
         try:
             claims = json.loads(token.claims)
         except Exception as e:
             raise errors.AuthenticationError(
                 f'authentication failed: malformed claims section in JWT'
             ) from None
 
-        if not claims.get(f"{namespace}.any_role"):
-            token_roles = claims.get(f"{namespace}.roles")
-            if not isinstance(token_roles, list):
-                raise errors.AuthenticationError(
-                    f'authentication failed: malformed claims section in JWT'
-                    f' expected mapping in "role_names"'
-                )
+        self._check_jwt_authz(claims, token_version, user)
+
+    def _check_jwt_authz(self, claims, token_version, user):
+        token_instances = None
+        token_roles = None
+        token_databases = None
+
+        if token_version == 1:
+            token_roles = self._get_jwt_edb_scope(claims, "edb.r")
+            token_instances = self._get_jwt_edb_scope(claims, "edb.i")
+            token_databases = self._get_jwt_edb_scope(claims, "edb.d")
+        else:
+            namespace = "edgedb.server"
+            if not claims.get(f"{namespace}.any_role"):
+                token_roles = claims.get(f"{namespace}.roles")
+                if not isinstance(token_roles, list):
+                    raise errors.AuthenticationError(
+                        f'authentication failed: malformed claims section in'
+                        f' JWT: expected a list in "{namespace}.roles"'
+                    )
+            else:
+                token_roles = None
+
+        if (
+            token_instances is not None
+            and self.server.get_instance_name() not in token_instances
```

Comment:
Similar (future) question as @tailhook 's, regardless of this PR tho: instance name is currently set without the org slug in the cloud for now, this allows token survive transferring instance among orgs, but not renaming the instance within the org. I noticed that instance name in the token can be a list (to support multiple instances of course) - this happens to allow creating new tokens to also support old servers lingering before the renaming is complete, but not old token on servers with new instance names after renaming. Q: do we want the tokens to be still valid after moving/renaming instances?

Reply:
> do we want the tokens to be still valid after moving/renaming instances?

I think in most cases instance names and tokens will either be stored together, or instance name would be harder to update than token (i.e. because instance name is in static environment, but token is in secrets). So I don't think we should optimize for that use case.

Reply:
Right, in that case maybe we should just add the full instance name into the token like `org_slug/inst_name` instead of just `inst_name` 🤔 (but this is a cloud thing, out of this PR)

Reply:
Yes, we'll use the full name.